File: jbake-core/src/main/java/org/jbake/parser/MarkupEngine.java
package org.jbake.parser;

import org.apache.commons.configuration.CompositeConfiguration;
import org.apache.commons.configuration.Configuration;
import org.apache.commons.io.IOUtils;
import org.jbake.app.Crawler;
import org.jbake.app.configuration.DefaultJBakeConfiguration;
import org.jbake.app.configuration.JBakeConfiguration;
import org.json.simple.JSONValue;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Map;

/**
 * Base class for markup engine wrappers. A markup engine is responsible for rendering
 * markup in a source file and exporting the result into the {@link ParserContext#getDocumentModel() contents} map.
 * <p>
 * This specific engine does nothing, meaning that the body is rendered as raw contents.
 *
 * @author Cédric Champeau
 */
public abstract class MarkupEngine implements ParserEngine {
    private static final Logger LOGGER = LoggerFactory.getLogger(MarkupEngine.class);
    private static final String UTF_8_BOM = "\uFEFF";

    private JBakeConfiguration configuration;

    /**
     * Tests if this markup engine can process the document.
     *
     * @param context the parser context
     * @return true if this markup engine has enough context to process this document. false otherwise
     */
    public boolean validate(ParserContext context) {
        return true;
    }

    /**
     * Processes the document header. Usually subclasses will parse the document body and look for
     * specific header metadata and export it into {@link ParserContext#getDocumentModel() contents} map.
     *
     * @param context the parser context
     */
    public void processHeader(final ParserContext context) {
    }

    /**
     * Processes the body of the document. Usually subclasses will parse the document body and render
     * it, exporting the result using the {@link org.jbake.parser.ParserContext#setBody(String)} method.
     *
     * @param context the parser context
     */
    public void processBody(final ParserContext context) {
    }

    @Override
    public Map<String, Object> parse(Configuration config, File file, String contentPath) {
        return parse(new DefaultJBakeConfiguration((CompositeConfiguration) config), file);
    }

    /**
     * Parse given file to extract as much infos as possible
     *
     * @param file file to process
     * @return a map containing all infos. Returning null indicates an error, even if an exception would be better.
     */
    public Map<String, Object> parse(JBakeConfiguration config, File file) {
        this.configuration = config;
        List<String> fileContents;
        try (InputStream is = new FileInputStream(file)) {

            fileContents = IOUtils.readLines(is, config.getRenderEncoding());
        } catch (IOException e) {
            LOGGER.error("Error while opening file {}", file, e);

            return null;
        }

        boolean hasHeader = hasHeader(fileContents);
        ParserContext context = new ParserContext(
                file,
                fileContents,
                config,
                hasHeader
        );

        if (hasHeader) {
            // read header from file
            processDefaultHeader(context);
        }
        // then read engine specific headers
        processHeader(context);

        setModelDefaultsIfNotSetInHeader(context);
        sanitizeTags(context);

        if (context.getType().isEmpty() || context.getStatus().isEmpty()) {
            // output error
            LOGGER.warn("Parsing skipped (missing type or status value in header meta data) for file {}!", file);
            return null;
        }

        // generate default body
        processDefaultBody(context);

        // eventually process body using specific engine
        if (validate(context)) {
            processBody(context);
        } else {
            LOGGER.error("Incomplete source file ({}) for markup engine: {}", file, getClass().getSimpleName());
            return null;
        }
        // TODO: post parsing plugins to hook in here?

        return context.getDocumentModel();
    }

    private void sanitizeTags(ParserContext context) {
        if (context.getTags() != null) {
            String[] tags = (String[]) context.getTags();
            for (int i = 0; i < tags.length; i++) {
                tags[i] = sanitizeValue(tags[i]);
                if (context.getConfig().getSanitizeTag()) {
                    tags[i] = tags[i].replace(" ", "-");
                }
            }
            context.setTags(tags);
        }
    }

    private void setModelDefaultsIfNotSetInHeader(ParserContext context) {
        if (context.getDate() == null) {
            context.setDate(new Date(context.getFile().lastModified()));
        }

        // default status has been set
        if (context.getConfig().getDefaultStatus() != null && context.getStatus().isEmpty()) {
            // file hasn't got status so use default
            context.setDefaultStatus();
        }

        // default type has been set
        if (context.getConfig().getDefaultType() != null && context.getType().isEmpty()) {
            // file hasn't got type so use default
            context.setDefaultType();
        }
    }

    /**
     * Checks if the file has a meta-data header.
     *
     * @param contents Contents of file
     * @return true if header exists, false if not
     */
    private boolean hasHeader(List<String> contents) {
        boolean headerValid = true;
        boolean statusFound = false;
        boolean typeFound = false;

        if (!headerSeparatorDemarcatesHeader(contents)) {
            return false;
        }

        for (String line : contents) {
            if (hasHeaderSeparator(line)) {
                LOGGER.debug("Header separator found");
                break;
            }
            if (isTypeProperty(line)) {
                LOGGER.debug("Type property found");
                typeFound = true;
            }

            if (isStatusProperty(line)) {
                LOGGER.debug("Status property found");
                statusFound = true;
            }
            if (!line.isEmpty() && !line.contains("=")) {
                LOGGER.error("Property found without assignment [{}]", line);
                headerValid = false;
            }
        }
        return headerValid && (statusFound || hasDefaultStatus()) && (typeFound || hasDefaultType());
    }

    private boolean hasDefaultType() {
        return !configuration.getDefaultType().isEmpty();
    }

    private boolean hasDefaultStatus() {
        return !configuration.getDefaultStatus().isEmpty();
    }

    private boolean hasHeaderSeparatorInContent(List<String> contents) {
        return contents.indexOf(configuration.getHeaderSeparator()) != -1;
    }

    /**
     * Checks if header separator demarcates end of metadata header
     *
     * @param contents
     * @return true if header separator resides at end of metadata header, false if not
     */
    private boolean headerSeparatorDemarcatesHeader(List<String> contents) {
        List<String> subContents = null;
        int index = contents.indexOf(configuration.getHeaderSeparator());
        if (index != -1) {
            // get every line above header separator
            subContents = contents.subList(0, index);

            for (String line : subContents) {
                // header should only contain empty lines or lines with '=' in
                if (!line.contains("=") && !line.isEmpty())  {
                    return false;
                }
            }
            return true;
        } else {
            return false;
        }
    }

    private boolean hasHeaderSeparator(String line) {
        return line.equals(configuration.getHeaderSeparator());
    }

    private boolean isStatusProperty(String line) {
        return line.startsWith("status=");
    }

    private boolean isTypeProperty(String line) {
        return line.startsWith("type=");
    }

    /**
     * Process the header of the file.
     *
     * @param context the parser context
     */
    private void processDefaultHeader(ParserContext context) {
        for (String line : context.getFileLines()) {

            if (hasHeaderSeparator(line)) {
                break;
            }
            processLine(line, context.getDocumentModel());
        }
    }

    private void processLine(String line, Map<String, Object> content) {
        String[] parts = line.split("=", 2);
        if (!line.isEmpty() && parts.length == 2) {


            String key = sanitizeKey(parts[0]);
            String value = sanitizeValue(parts[1]);

            if (key.equalsIgnoreCase(Crawler.Attributes.DATE)) {
                DateFormat df = new SimpleDateFormat(configuration.getDateFormat());
                try {
                    Date date = df.parse(value);
                    content.put(key, date);
                } catch (ParseException e) {
                    LOGGER.error("unable to parse date {}", value);
                }
            } else if (key.equalsIgnoreCase(Crawler.Attributes.TAGS)) {
                content.put(key, getTags(value));
            } else if (isJson(value)) {
                content.put(key, JSONValue.parse(value));
            } else {
                content.put(key, value);
            }
        }
    }

    private String sanitizeValue(String part) {
        return part.trim();
    }

    private String sanitizeKey(String part) {
        String key;
        if (part.contains(UTF_8_BOM)) {
            key = part.trim().replace(UTF_8_BOM, "");
        } else {
            key = part.trim();
        }
        return key;
    }

    private String[] getTags(String tagsPart) {
        String[] tags = tagsPart.split(",");
        for (int i = 0; i < tags.length; i++)
            tags[i] = sanitizeValue(tags[i]);
        return tags;
    }

    private boolean isJson(String part) {
        return part.startsWith("{") && part.endsWith("}");
    }

    /**
     * Process the body of the file.
     *
     * @param context the parser context
     */
    private void processDefaultBody(ParserContext context) {
        StringBuilder body = new StringBuilder();
        boolean inBody = false;
        for (String line : context.getFileLines()) {
            if (inBody) {
                body.append(line).append("\n");
            }
            if (line.equals(configuration.getHeaderSeparator())) {
                inBody = true;
            }
        }

        if (body.length() == 0) {
            for (String line : context.getFileLines()) {
                body.append(line).append("\n");
            }
        }
        context.setBody(body.toString());
    }
}

File: jbake-core/src/main/kotlin/org/jbake/parser/MarkupEngine.kt
package org.jbake.parser

import org.apache.commons.io.IOUtils
import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.model.DocumentModel
import org.jbake.model.ModelAttributes.DOC_DATE
import org.jbake.model.ModelAttributes.DOC_TAGS
import org.jbake.model.ModelAttributes.FS_DOC_IS_CACHED_IN_DB
import org.jbake.util.Logging.logger
import org.jbake.util.trace
import org.json.simple.JSONValue
import org.slf4j.Logger
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import kotlin.time.ExperimentalTime
import kotlin.time.Instant
import kotlin.time.toJavaInstant


/**
 * Base class for markup engine wrappers. A markup engine is responsible for rendering markup in a source file
 * and exporting the result into the [contents][ParserContext.documentModel] map.
 */
abstract class MarkupEngine : ParserEngine {

    private var configuration: JBakeConfiguration? = null

    /**
     * @return true if this markup engine has enough context to process this document. false otherwise
     */
    fun validate(context: ParserContext) = true

    /**
     * Processes the document header. Usually subclasses will parse the document body and look for
     * specific header metadata and export it into [contents][ParserContext.documentModel] map.
     */
    open fun processHeader(context: ParserContext) {}

    /**
     * Processes the body of the document. Usually, subclasses will parse the document body and render it,
     * exporting the result as [ParserContext.body].
     */
    open fun processBody(parserContext: ParserContext) {}

    /**
     * Parse the given file to extract as many infos as possible.
     * @return A map containing all infos. null indicates an error. TODO: Exception would be better.
     */
    override fun parse(config: JBakeConfiguration, file: File): DocumentModel? {
        this.configuration = config
        val fileContent = getFileContent(file, config.renderEncoding!!)

        if (fileContent.isEmpty()) return null

        val hasHeader = hasDocumentHeader(fileContent)
        val context = ParserContext(file, fileContent, config, hasHeader)

        // Read header from file.
        processDefaultHeader(context)
        // Then read engine specific headers.
        processHeader(context)

        setModelDefaultsIfNotSetInHeader(context)
        sanitizeTags(context)

        if (context.type.isEmpty() || context.status.isEmpty()) {
            log.warn("Parsing skipped - missing type or status value in header meta data for file: ${file.path}")
            return null
        }

        // Generate default body.
        processDefaultBody(context)

        if (!validate(context)) {
            log.error("Incomplete source file (${file.path}) for markup engine: ${javaClass.simpleName}")
            return null
        }
        // Eventually process body using specific engine.
        processBody(context)

        // TODO: post-parsing plugins to hook in here?
        return context.documentModel
    }

    private fun getFileContent(file: File, encoding: String): MutableList<String> {
        try {
            FileInputStream(file).use { `is` ->
                log.trace { "Read file '$file' with encoding '$encoding'" }
                val lines = IOUtils.readLines(`is`, encoding)
                if (!lines.isEmpty() && isUtf8WithBOM(encoding, lines[0])) {
                    log.warn("Removed BOM from file '$file' read with encoding '$encoding'")
                    lines[0] = lines[0].replace(UTF_8_BOM, "")
                }
                return lines
            }
        } catch (e: IOException) {
            log.error("Error while opening file $file", e)
            return mutableListOf()
        }
    }

    private fun isUtf8WithBOM(encoding: String, line: String): Boolean {
        return encoding == "UTF-8" && line.startsWith(UTF_8_BOM)
    }

    private fun sanitizeTags(context: ParserContext) {
        val tags = context.tags
            .map { sanitize(it) }
            .let { if (context.config.sanitizeTag) it.map { it.replace(" ", "-")  } else it }
        context.setTags(tags)
    }

    @OptIn(ExperimentalTime::class)
    private fun setModelDefaultsIfNotSetInHeader(context: ParserContext) {
        if (context.date == null)
            context.date =
                Instant.fromEpochMilliseconds(context.file.lastModified()).toJavaInstant()
                    .atZone(context.config.freemarkerTimeZone.toZoneId()).toOffsetDateTime()

        if (context.config.defaultStatus != null && context.status.isEmpty())
            context.setDefaultStatus()

        if (context.config.defaultType != null && context.type.isEmpty())
            context.setDefaultType()

        if (context.config.defaultAuthor != null)
            context.setDefaultAuthor()
    }

    /**
     * Checks if the file has a meta-data header.
     */
    private fun hasDocumentHeader(contents: MutableList<String>): Boolean {
        var headerValid = true
        var statusFound = false
        var typeFound = false

        if (!headerSeparatorDemarcatesHeader(contents)) return false

        for (line in contents) {
            if (hasHeaderSeparator(line)) { log.trace("Header separator found"); break }
            if (isTypeProperty(line)) { log.trace("Type property found"); typeFound = true }
            if (isStatusProperty(line)) { log.trace("Status property found"); statusFound = true }
            if (!line.isEmpty() && !line.contains("=")) {
                log.error("Property found without assignment [$line]")
                headerValid = false
            }
        }
        return headerValid && (statusFound || hasDefaultStatus()) && (typeFound || hasDefaultType())
    }

    private fun hasDefaultType() = !configuration!!.defaultType!!.isEmpty()

    private fun hasDefaultStatus() = !configuration!!.defaultStatus!!.isEmpty()


    /**
     * Checks if header separator demarcates end of metadata header.
     * @return true if header separator resides at end of metadata header, false if not
     */
    private fun headerSeparatorDemarcatesHeader(contents: MutableList<String>): Boolean {
        val index = contents.indexOf(configuration!!.headerSeparator)
        if (index == -1)
            return false

        // Get every line above header separator.
        return contents.subList(0, index).all { it.contains("=") || it.isEmpty() }
    }

    private fun hasHeaderSeparator(line: String)
        = sanitize(line) == configuration!!.headerSeparator

    private fun isStatusProperty(line: String)
        = sanitize(line).startsWith("status=")

    private fun isTypeProperty(line: String)
        = sanitize(line).startsWith("type=")

    /** Process the header of the file. */
    private fun processDefaultHeader(context: ParserContext) {
        if (!context.hasHeader()) return

        for (line in context.fileLines) {
            if (hasHeaderSeparator(line)) break
            processHeaderLine(line, context.documentModel)
        }
    }

    private fun processHeaderLine(line: String, content: DocumentModel) {
        val parts = line.split("=".toRegex(), limit = 2).toTypedArray()
        if (!line.isEmpty() && parts.size == 2)
            storeHeaderValue(parts[0], parts[1], content)
    }

    val formatter: DateTimeFormatter by lazy { configuration?.dateFormat?.let { DateTimeFormatter.ofPattern(it) } ?: DateTimeFormatter.ISO_LOCAL_DATE }

    fun storeHeaderValue(inputKey: String, inputValue: String, content: DocumentModel) {
        val key = sanitize(inputKey).lowercase()
        val value = sanitize(inputValue)

        when {
            key == DOC_DATE -> {
                runCatching {
                    // Try parsing as OffsetDateTime first (for full date-time with offset)
                    content.date = OffsetDateTime.parse(value, formatter)
                }
                .recoverCatching {
                    // If that fails, try parsing as LocalDate and convert to OffsetDateTime at start of day in system timezone
                    val localDate = java.time.LocalDate.parse(value, formatter)
                    content.date = localDate.atTime(0,0,0).atOffset(java.time.ZoneOffset.UTC)
                    // TODO: This was .atStartOfDay(), but that doesn't give seconds, so parsing failed. Investigate how to treat that later in processing.
                }
                .onFailure { e -> log.error("Unable to parse date $value with format '${configuration?.dateFormat}': ${e.message}") }
            }
            key == DOC_TAGS            -> content.tags = getTags(value)
            key == FS_DOC_IS_CACHED_IN_DB          -> content.cached = (value.toBoolean())
            isJson(value)    -> content[key] = JSONValue.parse(value)
            else                   -> content[key] = value
        }
    }

    private fun sanitize(part: String) = part.trim { it <= ' ' }

    private fun getTags(tagsPart: String): List<String>
        = tagsPart.split(",".toRegex()).dropLastWhile { it.isEmpty() }

    private fun isJson(part: String)
        = part.trim().let { it.startsWith("{") && it.endsWith("}") }


    /** Process the body of the file. */
    private fun processDefaultBody(context: ParserContext) {
        val headerSeparator = configuration!!.headerSeparator
        val body = context.fileLines
            .dropWhile { it != headerSeparator }
            .drop(1) // Skip the separator itself
            .joinToString("\n")
            .ifEmpty { context.fileLines.joinToString("\n") }

        context.body = body
    }

    private val UTF_8_BOM = "\uFEFF"
    private val log: Logger by logger()
}


/** This specific engine does nothing, meaning that the body is rendered as raw contents. */
class RawMarkupEngine : MarkupEngine()

/**
 * An internal rendering engine used to notify the user that the markup format he used requires an engine that couldn't be loaded.
 */
class ErrorEngine @JvmOverloads constructor(private val engineName: String = "unknown")
    : MarkupEngine()
{
    override fun processHeader(context: ParserContext) {
        val documentModel = context.documentModel
        documentModel.type = "post"
        documentModel.status = "published"
        documentModel.title = "Rendering engine missing"
        documentModel.date = OffsetDateTime.now().truncatedTo(ChronoUnit.SECONDS)
        documentModel.tags = emptyList()
    }

    override fun processBody(parserContext: ParserContext) {
        parserContext.body = "The markup engine [" + engineName + "] for [" + parserContext.file + "] couldn't be loaded"
    }
}
